### PR TITLE
re-resolve broken k_busy_wait on Nordic platforms

### LINKED ISSUE #11626: k_busy_wait exits early on Nordic chips
A `k_busy_wait(20)` in driver code to wait for a reset to complete was proceeding in 13 us, causing a hardware failure.  While investigating I discovered that for Nordic NRF51 k_busy_wait delays:

* 12.38 us for an argument of 1, 2, 5, 10, 15, or 20 us
* 26 to 51 us for an argument of 50 us
* 54 to 57 us for an argument of 75 us
* 83 to 86 us for an argument of 100 us

Further, for NRF52840 k_busy_wait delays:
* 3.9 to 4.58 us for an argument of 1, 2, 5, 10, 15, or 20 us
* 18 to 31 us for an argument of 50 us
* 61 us for an argument of 75 us
* 89 to 91 us for an argument of 100 us

Timings derived from a custom application with a GPIO toggle before and after the delay, duration measured with Saleae Logic Pro-8.

That really short delays take much longer than they should is unfortunate.  That longer delays take shorter than they should is simply unacceptable.  We must have a reliable usleep functionality to satisfy hardware timing requirements.

I did find issue #6498 which was enlightening but did not provide a solution.

Some improvement might be accomplished by fixing k_busy_wait to account for cycles lost to integer truncation, though a 10% error in a 100 us delay is difficult to understand.  However, I'm fixing it another way in a PR to be submitted next.

### PR DESCRIPTION
**Pull request by pabigot**
## soc: arm: nordic_nrf: change default SYS_CLOCKS_PER_SEC

The default system clock on all Nordic devices is based on a 32 KiHz
(2^15 Hz) timer.  Scheduling ticks requires that deadlines be specified
with a timer counter that aligns to a system clock.  With the Zephyr
default 100 clocks-per-sec configuration this results in 100 ticks every
32700 ticks of the cycle timer.  This reveals two problems:

* The uptime clock misrepresents elapsed time because it runs 0.208%
  (68/32768) faster than the best available clock;

* Calculation of timer counter compare values often requires an integer
  division and multiply operation to produce a value that's a multiple
  of clock-ticks-per-second.

Integer division on the Cortex-M1 nRF51 is done in software with a
(value-dependent) algorithm with a non-constant runtime that can be
significant.  This can produce missed Bluetooth deadlines as discussed
in upstream #14577 and others.

By changing the default divisor to one that evenly divides the 2^15
clock rate the time interrupts are disabled to manage timers is
significantly reduced, as is the error between uptime and real time.  Do
this at the top level, moving SYS_CLOCK_HW_CYCLES_PER_SEC there as well
since the two parameters are related.

Note that the central_hr configuration described in upstream #13610 does
not distinguish latency due to timer management from other
irq_block/spinlock regions, and the maximum observed latency will still
exceed the nominal 10 us allowed maximum.  However this does occur
much less frequently than changing the timer deadline which can happen
multiple times per tick.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

:100644 100644 00f3947f39 d0251bbc23 M	soc/arm/nordic_nrf/nrf51/Kconfig.defconfig.series
:100644 100644 cff0b9fad9 a91e5efc3f M	soc/arm/nordic_nrf/nrf52/Kconfig.defconfig.series
:100644 100644 d88da8aef2 68f636acc0 M	soc/arm/nordic_nrf/nrf91/Kconfig.defconfig.series

## soc: arm: nordic_nrf: unrevert provide custom busy_wait implementations

This reverts commit bd24b31139c1946bd9fe4c336c697b69608fac5e.

While the test case failure described in #14186 is associated with the
cycle-based busy-wait implementation, that test is fragile, and fails
less frequently once the incongruence between ticks-per-second and the
32 KiHz RTC clock are resolved.  It also assumes that the system clock
is more stable than the infrastructure underlying the the busy-wait
implementation, which is not necessarily true.

The gross inaccuracies in the standard busy-wait on Nordic described in
issue #11626 justify restoring the custom solution.

As this applies to all Nordic devices, move the setting to the top-level
Kconfig.defconfig.

See: https://github.com/zephyrproject-rtos/zephyr/issues/11626#issuecomment-487243369

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

:100644 100644 d0251bbc23 dd6ddbb486 M	soc/arm/nordic_nrf/nrf51/Kconfig.defconfig.series
:100644 100644 eb235d1cf6 cb3021a748 M	soc/arm/nordic_nrf/nrf51/soc.c
:100644 100644 a91e5efc3f 78a903d659 M	soc/arm/nordic_nrf/nrf52/Kconfig.defconfig.series
:100644 100644 44e9edbdf8 eb1f979602 M	soc/arm/nordic_nrf/nrf52/soc.c

NB: The reversion did not affect the use of the custom busywait in nrf91.

Closes #11626

### CODE REVIEW

**Comment by pabigot** [comment]
This isn't quite right; see further discussion in https://github.com/zephyrproject-rtos/zephyr/issues/11626#issuecomment-487243369.

In any case I believe the first patch in this PR is valuable and should be merged, for the reasons given in the commit message.

I personally consider the objective gross inaccuracies of busywait for short durations in #11626 to be far more worrisome than differences between CPU cycles and the system clock for 1 s delays where `k_sleep()` should have been used instead, so I'm taking the second patch in my fork.  In any case, merging or not is up to the Nordic folks.

If you chose not to unrevert the custom busywait you might want to complete the reversion for  nRF91 which still uses the custom one.

**Review by pizi-nordic — APPROVED** [review]

**Comment by carlescufi** [comment]
@pizi-nordic can you add a patch to this branch that completes the reversion for nRF91?

**Comment by carlescufi** [comment]
@pizi-nordic or @anangl can you add a patch to this branch that completes the reversion for nRF91?

**Comment by pabigot** [comment]
@carlescufi nRF91 was never reverted.  With this PR in this state all three targets are treated the same.

**Comment by pabigot** [comment]
I've updated the commit message in the second commit (and its copy in the PR description) to correctly reflect the conclusions related to #14186.

**Review by anangl — APPROVED** [review]

File: soc/arm/nordic_nrf/nrf51/Kconfig.defconfig.series [hunk 1]
```diff
@@ -17,6 +17,13 @@ config SYS_CLOCK_HW_CYCLES_PER_SEC
 	int
 	default 32768
 
+config SYS_CLOCK_TICKS_PER_SEC
```
Comment (ioannisg) [inline soc/arm/nordic_nrf/nrf51/Kconfig.defconfig.series]: Would it make sense to move this definition one level up, so we don't get to write it three times?
One can always overwrite it in project...

File: soc/arm/nordic_nrf/nrf51/Kconfig.defconfig.series [hunk 2]
```diff
@@ -17,6 +17,13 @@ config SYS_CLOCK_HW_CYCLES_PER_SEC
...
+	int
+	default 128
+
+config ARCH_HAS_CUSTOM_BUSY_WAIT
```
Comment (ioannisg) [inline soc/arm/nordic_nrf/nrf51/Kconfig.defconfig.series]: Same question applies to this one (now that all nRF SOC have custom busy wait)

**Review by ioannisg — COMMENTED** [review]
Seems fine, I left a couple of comments/questions.

**Comment by pabigot** [comment]
@ioannisg Your suggestion of refactoring Kconfig makes sense.  Assuming @pizi-nordic and @anangl who have approved this in its current form agree in principle I agree that change is appropriate.

Before spending more time on rework, though, I'd like clarification from @carlescufi whose [last comment](https://github.com/zephyrproject-rtos/zephyr/pull/15707#issuecomment-489092440)  could be taken as direction to drop the patch restoring the cycle-based custom busywait to nrf5 and instead complete the reversion of the custom busywait for nRF91 that was missed because nRF91 didn't exist in Zephyr when the commit that was subsequently reverted was originally merged.

**Comment by carlescufi** [comment]
@pabigot after discussion with other people I actually would like to get this in. Can you make the change suggested by @ioannisg and I will merge? thanks again!

**Comment by pabigot** [comment]
@carlescufi Moved all related configs to the top level, including the cycles-per-sec which should be co-located with the override for ticks-per-sec.  Revised the commit messages including in the PR description.

File: soc/arm/nordic_nrf/nrf51/Kconfig.defconfig.series (hunk 1 above)
Comment (pabigot) [reply soc/arm/nordic_nrf/nrf51/Kconfig.defconfig.series]: Done; thanks.

**Review by ioannisg — APPROVED** [review]
thanks

**Comment by carlescufi** [comment]
@pabigot thank you!

**Pull request merged**